src/analysis/souffle/operations.dl

```diff
@@ -27,28 +28,42 @@
 // regardless of arity. Useful for rules that apply to any operation.
 .type OperandList = [ operand: AccessPath, next: OperandList ]
 
-// Operations of any arity go in this relation.
-.decl operation(owner: Principal, operator: Operator, result: AccessPath, operandList: OperandList)
+// An Operation.
+.type Operation = [owner: Principal, operator: Operator, result: AccessPath,
```

Comment:
In most cases, there is little to no difference between a tuple of arguments and a record containing that tuple of values. The one benefit of records is when you need to match the whole record without introspecting into any of its contents. In that case, you can just refer to the whole record by one name rather than elaborating out each of its fields. You can see me doing this in the iterative case for populating `operationToPartialOperandLists`, the one case for `operationToOperands`, the iterative case for `operationToPartialAttributeLists`, and the population of `operationToAttribute`.

Reply:
With records, can you conditionally bind them? e.g., we can do the following with relations: 
`some_result(x, y, z): -some_condition(x), some_condition(y), operation(x, y, z)`

Can you do something similar when operation is a record?

Reply:
Yes. Imagine that `x`, `y`, and `z` are combined into a record `[x, y, z]` and that `some_result` and `operation` have this record type as their one parameter type. We can then rewrite the above rule as follows:

```Datalog
some_result([x, y, z]): -some_condition(x), some_condition(y), operation([x, y, z]).
```

We can easily deconstruct the record in a fashion similar to `auto [x, y, z] = ...;` declarations in C++. If, however, we had another rule that previously had to be written as:

```Datalog
some_other_result(x, y, z) :- operation(x, y, z).
```

we can now more concisely write the rule as:

```Datalog
some_other_result(op) :- operation(op).
```